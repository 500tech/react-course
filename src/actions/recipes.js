import * as consts from '../consts/action-types';

export const addRecipe = (id, title, description) => ({
  type: consts.ADD_RECIPE,
  payload: {
    id,
    title: title.trim(),
    description: description.trim()
  }
});

export const toggleRecipe = (id) => ({
  type: consts.TOGGLE_RECIPE,
  payload: id
});

export const fetchRecipes = () => ({
  type: consts.API,
  payload: {
    url: 'recipes.json',
    success: consts.SET_RECIPES
  }
});