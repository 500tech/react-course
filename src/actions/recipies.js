import * as actions from '../consts/action-types';

export const addRecipe = (title) => ({
  type: actions.ADD_RECIPE,
  title: title.trim()
});

export const toggleFavorite = (id) => ({
  type: actions.TOGGLE_FAVORITE,
  id
})