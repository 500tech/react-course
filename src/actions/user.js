import * as consts from '../consts/action-types';

export const fetchUser = () => ({
  type: consts.API,
  payload: {
    url: 'user.json',
    success: consts.SET_USER
  }
});