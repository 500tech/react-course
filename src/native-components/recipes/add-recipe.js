import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

class AddRecipe extends React.Component {
  render() {
    return (
      <View>
        <Text>Add recipe</Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 20
  },
  inputWrapper: {
    width: '100%'
  },
  inputContainer: {
    marginBottom: 30,
    width: '100%'
  },
  inputLabel: {
    color: '#0e749a',
    fontSize: 14,
    fontWeight: 'bold',
    marginBottom: 10
  },
  textInput: {
    borderColor: '#0e749a',
    borderWidth: 1,
    borderRadius: 4,
    paddingLeft: 10,
    paddingRight: 10,
    paddingTop: 5,
    paddingBottom: 5,
    width: '100%',
    fontSize: 14
  },
  button: {
    height: 50,
    width: 200,
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#0e749a'
  },
  buttonLabel: {
    fontSize: 14,
    color: 'white'
  }
});

export default AddRecipe;