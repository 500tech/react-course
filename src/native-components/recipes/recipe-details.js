import React from 'react';
import { TouchableOpacity, Text, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

const RecipeDetails = ({ recipe, unselectRecipe }) => (
  <TouchableOpacity style={ styles.container }
                    onPress={ unselectRecipe }>
    <Text style={ styles.title }>
      { recipe.title }
    </Text>

    <Text style={ styles.description }>
      { recipe.description }
    </Text>
  </TouchableOpacity>
);

RecipeDetails.propTypes = {
  recipe: PropTypes.object.isRequired,
  unselectRecipe: PropTypes.func.isRequired
};

const styles = StyleSheet.create({
  container: {
    padding: 20,
    alignItems: 'center'
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    width: '100%',
    textAlign: 'center',
    marginBottom: 20
  },
  description: {
    fontSize: 14,
    marginBottom: 50
  }
});

export default RecipeDetails