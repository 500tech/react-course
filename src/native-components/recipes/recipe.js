import React from 'react';
import { TouchableOpacity, View, Text, StyleSheet, Image } from 'react-native';

const Recipe = ({ recipe }) => {
  return (
    <TouchableOpacity style={ styles.container }>
      <View style={ styles.recipeWrapper }>
        <Image source={require('../../assets/dish.png')}
               style={ styles.image }/>

        <Text style={ styles.title }>
          { recipe.title }
        </Text>
      </View>

      <View style={ styles.separator }/>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: 50,
    alignItems: 'center',
    marginBottom: 20
  },
  recipeWrapper: {
    width: '100%',
    flexDirection: 'row'
  },
  image: {
    width: 30,
    height: 20,
    marginRight: 15
  },
  title: {
    fontWeight: '300',
    color: '#535353'
  },
  separator: {
    marginTop: 10,
    height: 1,
    backgroundColor: '#d2d2d2',
    width: '95%',
    alignSelf: 'flex-end'
  }
});

export default Recipe;