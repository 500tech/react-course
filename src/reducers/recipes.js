import * as consts from '../consts/action-types';

const initialState = [];

const reducer = (recipes = initialState, action) => {
  switch (action.type) {
    case consts.ADD_RECIPE:
      const newRecipe = Object.assign({}, action.payload, {
        favorite: false
      });

      return recipes.concat(newRecipe);

    case consts.TOGGLE_RECIPE:
      return recipes.map(recipe => recipe.id !== action.payload
        ? recipe
        : Object.assign({}, recipe, { favorite: !recipe.favorite }));

    case consts.SET_RECIPES:
      return action.payload;

    default:
      return recipes;
  }
};

export default reducer;