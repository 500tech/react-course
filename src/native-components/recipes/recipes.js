import React from 'react';
import { connect } from 'react-redux';
import { toggleRecipe, fetchRecipes } from '../../actions/recipes';
import { View, ScrollView, StyleSheet, RefreshControl } from 'react-native';

import Recipe from 'native/recipes/recipe';

class Recipes extends React.Component {
  state = {
    isLoading: false
  };

  componentDidMount() {
    this.props.fetchRecipes();
  }

  refreshList = () => {
    this.setState({ isLoading: true }, () => {
      this.props.fetchRecipes();

      this.setState({ isLoading: false });
    });
  };

  render() {
    const { recipes } = this.props;

    return (
      <View style={ styles.container }>
        <ScrollView style={{ width: '100%', flex: 1 }}
                    refreshControl={ <RefreshControl refreshing={ this.state.isLoading }
                                                     onRefresh={ this.refreshList }/> }>
          { recipes.map(recipe => <Recipe recipe={recipe} key={recipe.id}/>) }
        </ScrollView>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    paddingTop: 50,
    paddingBottom: 20,
    paddingLeft: 50,
    paddingRight: 50,
    backgroundColor: 'white',
    overflow: 'visible'
  },
  refreshButton: {
    marginTop: 10,
    height: 50,
    width: 200,
    borderRadius: 5,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#0e749a'
  },
  refreshButtonLabel: {
    fontSize: 14,
    color: 'white'
  }
});

const mapStateToProps = (state) => ({
  recipes: state.recipes
});

export default connect(mapStateToProps, {
  toggle: toggleRecipe,
  fetchRecipes
})(Recipes);