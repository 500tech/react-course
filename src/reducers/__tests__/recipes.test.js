import reducer from '../recipes';
import deepFreeze from 'deep-freeze';
import { addRecipe } from '../../actions/recipes';

const initialState = deepFreeze(reducer(undefined, { type: 'INIT' }));

describe('recipes reducer', () => {
  it('should handle unknown actions', () => {
    expect(reducer(initialState, { type: 'FAKE' })).toBe(initialState);
  });

  describe('add action', () => {
    it('should add recipe to empty list', () => {
      expect(reducer(initialState, addRecipe(1, 'test', 'desc'))).toMatchSnapshot()
    });

    it('should add recipe to a non-empty list', () => {
      const nonEmptyState = deepFreeze(reducer(initialState, addRecipe(1, 'first', '1')));

      expect(reducer(nonEmptyState, addRecipe(2, 'second', '2'))).toMatchSnapshot();
    });
  });
});