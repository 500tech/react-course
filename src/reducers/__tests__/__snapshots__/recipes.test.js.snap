// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipes reducer add action should add recipe to a non-empty list 1`] = `
Array [
  Object {
    "description": "1",
    "favorite": false,
    "id": 1,
    "title": "first",
  },
  Object {
    "description": "2",
    "favorite": false,
    "id": 2,
    "title": "second",
  },
]
`;

exports[`recipes reducer add action should add recipe to empty list 1`] = `
Array [
  Object {
    "description": "desc",
    "favorite": false,
    "id": 1,
    "title": "test",
  },
]
`;
